File: common/src/main/java/com/terrasect/common/handler/BiomeHandler.java
package com.terrasect.common.handler;

import com.terrasect.common.generation.MinecraftContext;
import com.terrasect.common.generation.World;
import com.terrasect.common.helpers.BiomeFilter;
import net.minecraft.core.Holder;
import net.minecraft.world.level.biome.Biome;
import net.minecraft.world.level.biome.Climate;

public final class BiomeHandler {

  private BiomeHandler() {}

  public static Holder<Biome> selectBiome(
      MinecraftContext context, int quartX, int quartZ, Climate.TargetPoint targetPoint) {

    var blockX = quartX << 2;
    var blockZ = quartZ << 2;

    var traversal = World.traverse(context, blockX, blockZ);
    var region = traversal != null ? traversal.region : null;

    if (region == null) {
      return selectFromBaseList(context, targetPoint);
    }

    var rules = region.definition().biomes();
    if (!BiomeFilter.hasRules(rules)) {
      return selectFromBaseList(context, targetPoint);
    }

    var filteredList = context.biomeLookup.getFilteredParameterList(rules);
    if (filteredList == null) {
      return selectFromBaseList(context, targetPoint);
    }

    return filteredList.findValue(targetPoint);
  }

  private static Holder<Biome> selectFromBaseList(
      MinecraftContext context, Climate.TargetPoint targetPoint) {
    return context.parameterList().findValue(targetPoint);
  }
}

File: common/src/main/kotlin/terrasect/handler/BiomeHandler.kt
package terrasect.handler

import net.minecraft.core.Holder
import net.minecraft.world.level.biome.Biome
import net.minecraft.world.level.biome.Climate
import terrasect.generation.DimensionContext
import terrasect.instrumentation.TerrasectInstr
import terrasect.instrumentation.TerrasectMetricEvent

private val instr = TerrasectInstr.biome
private val appliedCounter = instr.counter(TerrasectMetricEvent.BIOME_APPLIED)
private val rejectedCounter = instr.counter(TerrasectMetricEvent.BIOME_REJECTED)
private val fallbackCounter = instr.counter(TerrasectMetricEvent.BIOME_FALLBACK_APPLIED)
private val rejectedNoFallbackCounter =
  instr.counter(TerrasectMetricEvent.BIOME_REJECTED_NO_FALLBACK)

object BiomeHandler {
  @JvmStatic
  fun selectBiome(
    dimensionContext: DimensionContext?,
    quartX: Int,
    quartZ: Int,
    target: Climate.TargetPoint,
    base: Holder<Biome>,
  ): Holder<Biome> {
    val context = dimensionContext ?: return base
    val lookup = context.biomeLookup ?: return base
    val blockX = quartX shl 2
    val blockZ = quartZ shl 2
    val region = context.traverser.traverse(blockX, blockZ, context.cache).region
    if (!lookup.isConstrained(region)) return base

    appliedCounter.increment()
    if (lookup.isAdmitted(region, base.value())) return base

    rejectedCounter.increment()
    return lookup.select(region, target)?.also {
      fallbackCounter.increment()
    }
      ?: run {
        rejectedNoFallbackCounter.increment()
        base
      }
  }
}
